server/cluster/unsafe_recovery_controller.go

```diff
@@ -125,12 +128,216 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
+			return false
+		} else if target, ok := targetRegions[region.Id]; ok {
+			if bytes.Equal(target.StartKey, region.StartKey) && bytes.Equal(target.EndKey, region.EndKey) {
+				numFinished += 1
+			} else {
+				return false
+			}
+		}
+	}
+	return numFinished == len(targetRegions)
+
+}
+
+type regionItem struct {
+	region *metapb.Region
+}
+
+func (r regionItem) Less(other btree.Item) bool {
+	return bytes.Compare(r.region.StartKey, other.(regionItem).region.StartKey) < 0
+}
+
+func (u *unsafeRecoveryController) canElectLeader(region *metapb.Region) bool {
+	numFailedPeers := 0
+	for _, peer := range region.Peers {
+		if _, ok := u.failedStores[peer.StoreId]; ok {
+			numFailedPeers += 1
+		}
+	}
+	return numFailedPeers*2 < len(region.Peers)
+}
+
+func (u *unsafeRecoveryController) removeFailedStores(region *metapb.Region) bool {
+	updated := false
+	var newPeerList []*metapb.Peer
+	for _, peer := range region.Peers {
+		if _, ok := u.failedStores[peer.StoreId]; !ok {
+			newPeerList = append(newPeerList, peer)
+		} else {
+			updated = true
+		}
+	}
+	region.Peers = newPeerList
+	return updated
+}
+
+type peerStorePair struct {
+	peer    *pdpb.PeerReport
+	storeId uint64
 }
 
 func (u *unsafeRecoveryController) generateRecoveryPlan() {
 	u.Lock()
 	defer u.Unlock()
+	healthyRegions := make(map[uint64]*pdpb.PeerReport)
+	var allPeerReports []*peerStorePair
+	for storeId, storeReport := range u.storeReports {
+		for _, peerReport := range storeReport.Reports {
+			allPeerReports = append(allPeerReports, &peerStorePair{peerReport, storeId})
+			regionId := peerReport.RegionState.Region.Id
+			if existing, ok := healthyRegions[regionId]; ok && existing.RaftState.LastIndex > peerReport.RaftState.LastIndex {
+				// Only keeps the newest region report.
+				continue
+			}
+			if u.canElectLeader(peerReport.RegionState.Region) {
+				healthyRegions[regionId] = peerReport
+			} else {
+				delete(healthyRegions, regionId)
+			}
+		}
+	}
+	recoveredRanges := btree.New(2)
+	for _, report := range healthyRegions {
+		recoveredRanges.ReplaceOrInsert(regionItem{report.RegionState.Region})
+	}
+	sort.SliceStable(allPeerReports, func(i, j int) bool {
+		return allPeerReports[i].peer.RegionState.Region.RegionEpoch.Version > allPeerReports[j].peer.RegionState.Region.RegionEpoch.Version
+	})
+	for _, peerStorePair := range allPeerReports {
+		region := peerStorePair.peer.RegionState.Region
+		storeId := peerStorePair.storeId
+		updated := u.removeFailedStores(region)
+		// Iterates all leaderless peers, if a peer has been fully
+		// covered by a valid range (from another region or the same region but another peer), do nothing, while, if a peer can fill some
+		// of the gaps between available ranges so far, find all the gaps and ask the peer
+		// to create new regions for each of the gap, besides, merge the all the scattered
+		// ranges that are connected by filling up these gaps.
+		overlapRanges := make([]*metapb.Region, 0)
```

Comment:
can we extract a function called `getOverlaps`? the `overlapRanges` is a potential overlap regions list which is really confusing.

Reply:
Done